design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,214 @@
+# Monitoring code refactor
+
+**Authors:** João Vilaça <jvilaca@redhat.com>
+
+## What
+
+This design document is proposing a code refactor for the monitoring logic in
+all KubeVirt components. The goal is to have a centralized place for the
+creation of monitoring resources, a consistent monitoring package between all
+projects, a code struct that is easy to maintain and evolve, while moving closer
+to the Kubernetes metric implementation style.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+Right now, every change regarding monitoring needs to happen at many different
+places. For example, when adding the linter for metrics names, it was necessary
+to go to all KubeVirt components repositories and implement the linter
+validation in each. This is demanding, very time-consuming, requires going
+through multiple duplicated code reviews, uses many CI resources, and makes it
+mentally harder to decide to change anything.
+
+### Pitfalls of the current solution
+
+We are facing multiple challenges with the current solution:
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
+
+- Complex to collect all existing metrics and alerts for automatic documentation
+
+- High levels of intertwined monitoring and business logic code
+
+### Goals
+
+- Decouple monitoring logic from business logic
+- Have a centralized place for the configuration of monitoring resources
+- Encapsulate the monitoring best-practices and the common patterns into a
+library and have it as a dependency for all KubeVirt components
+- Have a declarative way to define metrics and alerts
```

Comment:
The current way is already declarative, is it not?

Reply:
not really, we want to abstract all the calls to `prometheus.MustRegister`, `prometheus.NewCounter`,  `prometheus.NewCounterVec`, and so on, all of that should be handled by the common lib

Reply:
I think the idea is, for metrics, to add the metrics to a list where each item includes all the required details and the list would be used for defining the based on its typpe, registering it, adding it to the auto docs generator and so on. Am I correct?

Reply:
@sradco @iholder101 I think the approach here is different for metrics and for alerts. With regards to metrics, are we talking about creating a code generator here?